src/rpc/mining.cpp

```diff
@@ -471,15 +484,17 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
+        RPCServer::BlockChangeBlocker block_waiter;
+
         // Release the wallet and main lock while waiting
         LEAVE_CRITICAL_SECTION(cs_main);
         {
             checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
 
-            WaitableLock lock(csBestBlock);
-            while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
+            WaitableLock lock(block_waiter.m_cs);
+            while (block_waiter.m_last_block_hash == hashWatchedChain && IsRPCRunning())
```

Comment:
In commit "Replace csBestBlock/cvBlockChanged with validationinterface "

IsRPCRunning does not appear to be thread safe. Maybe fRPCRunning be a changed to an atomic bool.